Source/Section/PDFSectionObject.swift

```diff
@@ -0,0 +1,153 @@
+			columnWidthSum += column.width
+
+			let columnLeftMargin = columnIndex == 0 ? 0 : section.columnMargin / 2
+			let columnRightMargin = columnIndex == section.columns.count - 1 ? 0 : section.columnMargin / 2
+
+			for container in [PDFContainer.contentLeft, .contentCenter, .contentRight] {
+				generator.setContentOffset(in: container, to: originalContentOffset)
+				generator.layout.indentation.setLeft(indentation: indentationLeft + columnLeftMargin, in: container)
+				generator.layout.indentation.setRight(indentation: generator.document.layout.contentSize.width - columnWidthSum * generator.document.layout.contentSize.width + columnRightMargin, in: container)
+			}
+
+			let object = PDFSectionColumnObject(column: column)
+			objectsPerColumn[columnIndex] = try object.calculate(generator: generator, container: container)
+
+			indentationLeft += column.width * generator.document.layout.contentSize.width
+		}
+		result += calulatePageBreakPositions(objectsPerColumn)
+		generator.layout.indentation.content = originalIndent
+
+		var contentMinY: CGFloat? = nil
+		var contentMaxY: CGFloat? = nil
+
+		for current in result.reversed() {
+			let currentObject = current.1
+
+			if currentObject is PDFPageBreakObject {
+				break
+			}
+
+			if contentMaxY == nil {
+				contentMaxY = currentObject.frame.maxY
+			} else if let maxY = contentMaxY, maxY < currentObject.frame.maxY {
+				contentMaxY = currentObject.frame.maxY
+			}
+
+			if contentMinY == nil {
+				contentMinY = currentObject.frame.minY
+			} else if let minY = contentMaxY, minY > currentObject.frame.minY {
+				contentMinY = currentObject.frame.minY
+			}
+		}
+		generator.setContentOffset(in: container, to: (contentMaxY ?? 0) - (contentMinY ?? 0))
+
+		return result
+	}
+	
+	/** The `PDFDocument` render engine calculates each object, which returns a list of calculated objects.
+	As an example if you add a text object, it will be calculated and return one text object which will then be rendered.
+	
+	**BUT** if the text is too long to fit the space, then it will be split up into two text objects with a `PDFPageBreakObject` in-between.
+	
+	During the render process whenever a page break object is found, it will create a new pdf page and continue there.
+	In order to render multi columns correctly, we need to merge the page breaks of all columns and make sure the page break occurs at the right time:
+	
+	```
+	All objects of column 1 before the first pagebreak
+	All objects of column 2 before the first pagebreak
+	All objects of column 3 before the first pagebreak
+	Pagebreak
+	All objects of column 1 after the first pagebreak up to the next pagebreak
+	All objects of column 2 after the first pagebreak up to the next pagebreak
+	All objects of column 3 after the first pagebreak up to the next pagebreak
+	Pagebreak
+	...
+	```
+	*/
+	func calulatePageBreakPositions(_ objectsPerColumn: [Int: [(PDFContainer, PDFObject)]]) -> [(PDFContainer, PDFObject)] {
```

Comment:
I am still not entirely sure what exactly this method does. Can you please add some comments what exactly this algorithm does?

Reply:
Improved the documentation with inline comments. Hoping this helps understanding how exactly the algorithm works.